site/en/blog/layoutng/index.md

```diff
@@ -0,0 +1,338 @@
+for inline layout, but also features paragraph-level caching for faster relayout, 
+shape-per-paragraph to apply font features across elements and words, 
+a new Unicode bidirectional algorithm using ICU, a lot of correctness fixes, and more.
+
+## Types of layout bugs
+
+Layout bugs broadly speaking fall into four different categories, 
+each with different root causes.
+
+### Correctness
+
+When we think about bugs in the rendering system, we typically think about correctness, 
+for example: "Browser A has X behaviour, while Browser B has Y behaviour", 
+or "Browsers A and B are both broken". 
+Previously this is what we spent a lot of our time on, 
+and in the process we were constantly fighting with the system. 
+A common failure mode was to apply a very targeted fix for one bug, 
+but find weeks later that we had caused a regression in another (seemingly unrelated) part of the system.
+
+As described in [previous posts](/blog/renderingng/#reliability), 
+this is a sign of a very brittle system. 
+For layout specifically, we didn't have a clean contract between any classes, 
+causing browser engineers to depend on state which they shouldn't, 
+or mis-interpret some value from another part of the system.
+
+As an example, at one point we had a chain of approximately 10 bugs over the course of more than a year, 
+related to flex layout. 
+Each fix caused either a correctness or performance issue in part of the system, 
+leading to yet another bug.
+
+Now that LayoutNG clearly defines the contract between all the components in the layout system, 
+we've found that we can apply changes with much more confidence. 
+We also benefit greatly from the excellent [Web Platform Tests](/blog/renderingng/#testing-and-metrics) (WPT) project, 
+which allows multiple parties to contribute to a common web test suite.
+
+Today we find that if we release a real regression on our stable channel, 
+it typically has no  associated tests in the WPT repository, 
+and doesn't result from a misunderstanding of component contracts. 
+Further, as part of our bugfix policy, we always add a new WPT test, 
+helping to ensure that no browser should make the same mistake again.
+
+### Under-invalidation
+
+If you've ever had a mysterious bug where resizing the browser window or toggling a CSS property magically makes the bug go away, 
+you've run into an under-invalidation problem. 
+Effectively a portion of the mutable tree was deemed clean, 
+but due to some change in parent constraints it didn't represent the correct output.
+
+This is very common with the two-pass 
+(walking the layout tree twice to determine the final layout state) layout modes described below. 
+Previously our code would look like:
+
+```cpp
+if (/* some very complicated statement */) {
+  child->ForceLayout();
+}
+
+A fix for this type of bug would typically be:
+
+if (/* some very complicated statement */ ||
+    /* another very complicated statement */) {
+  child->ForceLayout();
+}
+```
```

Comment:
This block might be better to split up into two, e.g.

```cpp
if (/* some very complicated statement */) {
  child->ForceLayout();
}
```

A fix for this type of bug would typically be:

```cpp
if (/* some very complicated statement */ ||
    /* another very complicated statement */) {
  child->ForceLayout();
}
```